File: structural/decorator.ts
// Component
interface Coffee {
  cost(): number;
  description(): string;
}

// Concrete Component
class BasicCoffee implements Coffee {
  cost(): number {
    return 5;
  }

  description(): string {
    return "Basic coffee";
  }
}

// Decorator
class CoffeeDecorator implements Coffee {
  protected coffee: Coffee;

  constructor(coffee: Coffee) {
    this.coffee = coffee;
  }

  cost(): number {
    return this.coffee.cost();
  }

  description(): string {
    return this.coffee.description();
  }
}

// Concrete Decorators
class Milk extends CoffeeDecorator {
  constructor(coffee: Coffee) {
    super(coffee);
  }

  cost(): number {
    return super.cost() + 1;
  }

  description(): string {
    return super.description() + ", Milk";
  }
}

class Sugar extends CoffeeDecorator {
  constructor(coffee: Coffee) {
    super(coffee);
  }

  cost(): number {
    return super.cost() + 0.5;
  }

  description(): string {
    return super.description() + ", Sugar";
  }
}

class Caramel extends CoffeeDecorator {
  constructor(coffee: Coffee) {
    super(coffee);
  }

  cost(): number {
    return super.cost() + 1.5;
  }

  description(): string {
    return super.description() + ", Caramel";
  }
}

// Example usage
let coffee: Coffee = new BasicCoffee();
console.log(`${coffee.description()} costs $${coffee.cost()}`);

coffee = new Milk(coffee);
console.log(`${coffee.description()} costs $${coffee.cost()}`);

coffee = new Sugar(coffee);
console.log(`${coffee.description()} costs $${coffee.cost()}`);

coffee = new Caramel(coffee);
console.log(`${coffee.description()} costs $${coffee.cost()}`);

File: structural/decorator.py
from typing import Protocol


# Component
class Coffee(Protocol):
    def cost(self) -> float: ...

    def description(self) -> str: ...


# Concrete Component
class BasicCoffee(Coffee):
    def cost(self) -> float:
        return 5.0

    def description(self) -> str:
        return "Basic coffee"


# Decorator
class CoffeeDecorator(Coffee):
    def __init__(self, coffee: Coffee) -> None:
        self._coffee = coffee

    def cost(self) -> float:
        return self._coffee.cost()

    def description(self) -> str:
        return self._coffee.description()


# Concrete Decorators
class Milk(CoffeeDecorator):
    def cost(self) -> float:
        return super().cost() + 1.0

    def description(self) -> str:
        return super().description() + ", Milk"


class Sugar(CoffeeDecorator):
    def cost(self) -> float:
        return super().cost() + 0.5

    def description(self) -> str:
        return super().description() + ", Sugar"


class Caramel(CoffeeDecorator):
    def cost(self) -> float:
        return super().cost() + 1.5

    def description(self) -> str:
        return super().description() + ", Caramel"


# Example usage
coffee: Coffee = BasicCoffee()
print(f"{coffee.description()} costs ${coffee.cost()}")

coffee = Milk(coffee)
print(f"{coffee.description()} costs ${coffee.cost()}")

coffee = Sugar(coffee)
print(f"{coffee.description()} costs ${coffee.cost()}")

coffee = Caramel(coffee)
print(f"{coffee.description()} costs ${coffee.cost()}")
